pandas/core/frame.py

```diff
@@ -3973,10 +3973,28 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         -------
         dataframe : DataFrame
         """
-        inplace = validate_bool_kwarg(inplace, 'inplace')
+        from pandas import Series
+
         if not isinstance(keys, list):
             keys = [keys]
 
+        # collect elements from "keys" that are not allowed array types
+        col_labels = [x for x in keys
+                      if not isinstance(x, (Series, Index, MultiIndex,
+                                            list, np.ndarray))]
+        if any(x not in self for x in col_labels):
+            # if there are any labels that are invalid, we raise a KeyError
+            missing = [x for x in col_labels if x not in self]
+            raise KeyError('{}'.format(missing))
+        elif len(set(col_labels)) < len(col_labels):
+            # if all are valid labels, but there are duplicates
+            dup = Series(col_labels)
```

Comment:
rather use a set difference operation here

Reply:
maybe I'm not seeing it, but set difference isn't gonna show duplicates (because as a set, they'll be the same)...

Reply:
subtract the sets of all - others

Reply:
this is not about the set of all inputs - it's strictly about column labels. E.g.
```
df.set_index(['A', 'A', 'B', df.A, some_series, some_index])
[...]
col_labels = ['A', 'A', 'B']  # at the line of your comment (in this case)
```
No set operation that I can think of would yield `A`, which is what I want to raise. (multisets would work, but that's hardly more intuitive than `duplicated`, IMO)